source/sessions/driver-sessions.rst

```diff
@@ -1113,16 +1141,20 @@ ensure that they close any explicit client sessions and any unexhausted cursors.
     * Create a MongoClient with the following options: ``maxPoolSize=1`` and ``retryWrites=true``
     * Attach a command started listener that collects each command's lsid
     * Initiate the following concurrent operations
-      * insertOne
-      * deleteOne
-      * updateOne
-      * bulkWrite ``[ { updateOne } ]``
-      * findOneAndDelete
-      * findOneAndUpdate
-      * findOneAndReplace
-      * find
+
+      * ``insertOne({ }),``
+      * ``deleteOne({ }),``
+      * ``updateOne({ }, { $set: { a: 1 } }),``
+      * ``bulkWrite([{ updateOne: { filter: { }, update: { $set: { a: 1 } } } }]),``
+      * ``findOneAndDelete({ }),``
+      * ``findOneAndUpdate({ }, { $set: { a: 1 } }),``
+      * ``findOneAndReplace({ }, { a: 1 }),``
+      * ``find().toArray()``
+
     * Wait for all operations to complete
-    * Assert that all commands contain the same lsid
+    * Drivers MAY assert that exactly one session is used for all the concurrent operations listed, however this can be nondeterministic if the session isn't released before check in. Drivers SHOULD NOT attempt to release before check in.
+    * Drivers SHOULD assert that after repeated runs they are able to achieve the use of exactly one session, this will statistically prove we've reduced the allocation amount
+    * Drivers MUST assert that the number of allocated sessions is strictly less than the number of concurrent operations. In this instance it would less than (but NOT equal to) 8.
```

Comment:
What do you think about:

```
    * Wait for all operations to complete successfully
    * Assert that all commands contain the same lsid. Note that it's possible, although very rare, for >1 server session to be used because the session is not released until after the connection is checked in. If this assertion fails, repeat this test 5 times until the assertion holds. 
```

Reply:
I added this phrasing / bullet but also kept what I had there are you thinking its overly verbose? I think it translates to actual code assertions well.

Reply:
I think you should make the "Drivers MAY assert that exactly one session is used..." part a sub-bullet of the bullet above, since it clarifies the context, and then you can get rid of the SHOULD assert bullet point about repeated runs (but definitely keep the last bullet point about the strictly less part - it doesn't hurt, and it's a good sanity check)

Reply:
I was suggesting to replace these 3 lines:
```
    * Drivers MAY assert that exactly one session is used for all the concurrent operations listed, however this can be nondeterministic if the session isn't released before check in. Drivers SHOULD NOT attempt to release before check in.
    * Drivers SHOULD assert that after repeated runs they are able to achieve the use of exactly one session, this will statistically prove we've reduced the allocation amount
    * Drivers MUST assert that the number of allocated sessions is strictly less than the number of concurrent operations. In this instance it would less than (but NOT equal to) 8.
```
I think it's a bit too verbose, for example the "Drivers SHOULD NOT attempt to release before check in." line doesn't really fit. I also believe the "strictly less than the number of concurrent operations" bullet is unneeded and confusing since we're already asserting that (eventually) only 1 session is used.

Reply:
The difference between "eventually one session is used" and "on every run the number of sessions is less than the number of concurrent operations" is that the latter guards against the unlikely but not impossible scenario of a driver having a race condition resulting in sometimes creating too many sessions. If we only keep your suggested phrasing, then the test would pass if we sometimes create one session and other times 16, and while we hope that no driver has such faulty logic, concurrent operations create the opportunity for all sorts of race conditions, so I think it's worth keeping a deterministic assertion that costs nothing to implement as a sanity check. 

But we certainly can and should update the phrasing to clarify that *this* assertion must be performed on every iteration, whereas the "exactly one" assertion just needs to happen once over the course of 5 tries.

Reply:
Okay I relent.

Reply:
Updated the language